### PR TITLE
fix assertion DFlash2 with `--split-mode tensor` on CUDA

### PR DESCRIPTION
**Pull request by art-den**
## Overview

Assertion is failed on llama-server startup if using dflash2 draft model and `--split-mode tensor` flag:

```
ggml\src\ggml-backend-meta.cpp:543: GGML_ASSERT(src_ss[0].axis != GGML_BACKEND_SPLIT_AXIS_0) failed
```

The changes in this PR address two issues:
- Crash in GGML_ASSERT (root cause)
- Reducing the VRAM footprint of the DFlash2 model for **`--split-mode tensor`**

## Additional information

Issue for this case: https://github.com/ggml-org/llama.cpp/issues/27819

### Root cause

The DFlash2 candidate selector runs `ggml_top_k()` on the lm_head output (`t_logits`) inside the graph. Under tensor parallelism `output.weight` is split along the vocab axis, so `t_logits` is split along axis 0, which `GGML_OP_TOP_K` (a per-row op) cannot handle in the meta backend.

**Fix:** When a DFlash2 draft is used with tensor parallelism, the lm_head must be replicated in full on every device (`GGML_BACKEND_SPLIT_AXIS_MIRRORED`) so the whole vocabulary is available in-graph. 

### Reducing the VRAM footprint

DFlash2's in-graph selector (`ggml_top_k` over the lm_head output) cannot run on a vocab-split lm_head under `--split-mode tensor`. Rather than replicating the lm_head on every device (MIRRORED, ~+500 MiB/device), the **selector now runs on the CPU from the gathered logits** + decoder hidden states, so the lm_head stays split.

VRAM savings for `https://huggingface.co/z-lab/Qwen3.8-27B-DFlash2-GGUF/blob/main/Qwen3.8-27B-DFlash2-Q4_K_M.gguf`:
- lm_head stays vocab-split instead of replicated: ~-500 MiB/device.
- Selector codebook tables are `TENSOR_SKIP`'d under tensor parallelism and read from the GGUF at init: ~-68 MiB/device.
- Draft ubatch is capped (floor 64) so the reserved `[vocab x ubatch]` compute buffer drops from ~539 MiB to ~67 MiB/device.

Without these modifications, the amount of occupied VRAM was x4.4 of the gguf file size.

NOTE: **No `ggml/` changes; the non-tensor path (single GPU / layer split) keeps the original in-graph selector.**

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure:  YES, Deepseek 4 Flash was used for investigation, fixing and reviewing

### CODE REVIEW

**Comment by art-den** [comment]
Bad news. VRAM usage for draft model is **x4.4** of original `Qwen3.8-27B-DFlash2-Q4_K_M.gguf` size for my two RTX 5060 Ti 16Gb. I have no idea how to fix it without large changes in sources.

I'm not sure yet if this MR is necessary ((

**Comment by mb1986** [comment]
> Bad news. VRAM usage for draft model is **x4.4** of original `Qwen3.8-27B-DFlash2-Q4_K_M.gguf` size for my two RTX 5060 Ti 16Gb. I have no idea how to fix it without large changes in sources.
> 
> I'm not sure yet if this MR is necessary ((

Have you tried to set `--parallel 1`?

It is possible for me to run `--split-mode tensor --spec-type draft-dflash --parallel 1` on two P40s.

b10679 + #25592 + #26636 + #27858

**Comment by art-den** [comment]
I think it's ready. I managed to reduce video memory consumption without a much rewriting. 

The current implementation produces **~73 t/s with dflash2** vs **~63 t/s with MTP** for my x2 RTX 5060 Ti 16Gb. 

Please check that I haven't broken dflash2 and dspark in single-GPU mode.

**Comment by pwilkin** [comment]
@ruixiang63 could you maybe take a look at this one? The diagnosis seems correct, but the code could use a look from a specdec expert.

**Comment by ruixiang63** [comment]
I don't have a good feeling about this PR, there are too much unnecessary changes. 

> Fix: When a DFlash2/DSpark draft is used with tensor parallelism, the lm_head must be replicated in full on every device (GGML_BACKEND_SPLIT_AXIS_MIRRORED) so the whole vocabulary is available in-graph.

This is the fix proposed in the description but there are so much irrlevant changes in the PR. Could you limit this PR to the changes required for the fix?

**Comment by art-den** [comment]
@ruixiang63 

I also have a bad feeling about my PR. I asked the agent to make as few changes as possible, but they ended up making far more than I expected.

The minimum change that would solve the problem is the first commit. But it has some unresolved issues (too much VRAM usage for the draft model).

This PR needs to be reviewed by someone familiar with llama.cpp. If these changes don't solve the problem correctly, I'll simply close the PR.

**Comment by ruixiang63** [comment]
The PR authors should be responsible for validating their own changes before requesting review. If you’re not confident in the implementation, please investigate and self-review it first. Asking an agent to write the code and then leaving reviewers or maintainers to determine whether it’s correct doesn't make any sense.

**Comment by art-den** [comment]
@ruixiang63, @pwilkin new changes for this PR:
- Updated comment for PR
- Fixed build for GCC
- Fixed comments from review bot
- Restored the non-tensor path (single GPU / layer split) to minimize PR changes
- Reverted fix for DSpark to minimize PR changes

**Comment by art-den** [comment]
Hm... `--no-mmproj-offload` is broken in this PR if DFlash2 used

**Comment by AiwendilInTheWoods** [comment]
Tested this branch on ROCm — it fixes the assert there too.

**Environment:** 2× AMD Radeon AI PRO R9700 (gfx1201, 32 GiB each), ROCm 7.2.3, kernel 6.8, `HIP_VISIBLE_DEVICES=0,1`. Branch `fix/dflash2-tensor-split` @ `3286b033c` built with `-DGGML_HIP=ON -DAMDGPU_TARGETS=gfx1201 -DCMAKE_HIP_ARCHITECTURES=gfx1201 -DGGML_HIP_GRAPHS=ON -DCMAKE_BUILD_TYPE=Release` (`/opt/rocm/llvm/bin/clang` 22). Target `unsloth/Qwen3.8-27B-GGUF` `Qwen3.8-27B-UD-Q6_K_XL.gguf`; drafter `z-lab/Qwen3.8-27B-DFlash2-GGUF` `Qwen3.8-27B-DFlash2-Q8_0.gguf` (the reconverted one with `dflash.rope.dimension_sections`).

**Before (master-based build without this branch):** at `n_max` 3, 5 and 7 under `--split-mode tensor`:
```
common_speculative_impl_draft_dflash: - n_max=7, n_min=0, p_min=0.00
common_speculative_impl_draft_dflash: - block_size=8, mask_token_id=248070, n_extract=5, sample_from_anchor=true
ggml/src/ggml-backend-meta.cpp:543: GGML_ASSERT(src_ss[0].axis != GGML_BACKEND_SPLIT_AXIS_0) failed
```

**With this branch:** loads and runs at all three depths; ~15 loads, no NaNs, no crashes.

```
llama-server -m Qwen3.8-27B-UD-Q6_K_XL.gguf -md Qwen3.8-27B-DFlash2-Q8_0.gguf \
  --spec-type draft-dflash --spec-draft-n-max 3 -sm tensor -fa on -c 131072 -ctk q8_0 -ctv q8_0 \
  -ngl 999 --jinja --parallel 1 --no-context-shift --host 127.0.0.1 --port 8091
```

**Results** (greedy, `temperature 0`, 4 distinct ~30-token prose prompts × 256 completion tokens × 3 reps, fresh server per config, same binary):

| config | tok/s (grand median) | mean accepted len | VRAM by PID, 2 GPUs |
|---|---|---|---|
| no speculation | 30.5 | — | 30.8 GB |
| `draft-mtp` n3 | 56.5 | 2.4–2.9 | 32.7 GB |
| `draft-dflash` n3 | 53.3 | 2.2–3.0 | 33.9 GB |
| `draft-dflash` n5 | 51.0 | 2.5–3.2 | 34.2 GB |
| `draft-dflash` n7 | 45.7 | 2.5–3.3 | 34.5 GB |

Output is deterministic across reps and byte-identical (SHA-256 of the completion) at n3, n5 and n7, and identical to `draft-mtp`'s on all four prompts. So the CPU-side selector is producing the same tokens as the target here.

`draft acceptance` in the slot timings: 0.41–0.65 at n3, 0.21–0.33 at n7 — accepted length doesn't grow with depth on this pairing, which is why n7 is slowest; that's a separate topic from this PR.

Happy to re-run anything useful (f16 KV, other depths, `--no-mmproj-offload`). Result JSON and server logs available.
